Adding More Changes to API

apanzerj opened the pull request:
Description:
- Added the attribute `template` to the Stage model.
- Modified StagesController (both) to allow setting the template stage.
- Nix the permalinks in the API

/cc @zendesk/samson
### Tasks
- [ ] :+1: from team
### References
- Jira link:
### Risks
- Level: Low/Med/High

sandlerr commented on app/views/shared/_header.html.erb:
```diff
@@ -74,8 +74,10 @@
           <li><%= link_to "Reports", csv_exports_path %></li>
-          <li><%= link_to "OAuth Applications", oauth_applications_path %></li>
-          <li><%= link_to "OAuth Authorized Applications", oauth_authorized_applications_path %></li>
+          <% if current_user.try(:super_admin?) %>
```
ruby 2.3 means you can use the lonely operator now :D `current_user&.super_admin?`

apanzerj replied:
wuuuuuuut.

crazy, but I like it.

sandlerr commented:
nitpick: would you mind pulling all the routing tests into a single block at the top of the file since that will map better to the code they're actually testing (i.e. the stuff in `routes.rb` rather than the actual actions)?

apanzerj commented on test/controllers/api/deploys_controller_test.rb:
```diff
@@ -3,17 +3,17 @@
 SingleCov.covered!
 
 describe Api::DeploysController do
+  assert_route "/api/deploys/active_count", 'api/deploys', 'active_count'
```
@sandlerr like this?

sandlerr replied:
yes! one more nitpick though, I would nest it inside 

```ruby
describe 'routing' do
  it 'dispatches correctly' do
    ...
  end
end
```

not sure what minitest's behaviour is if a spec fails outside of an `it` block, maybe it's fine.

sandlerr replied:
never mind, assert_route is an `it` block

apanzerj replied:
actually got it done anyways. sorta...
```ruby
def self.assert_route(verb:, path:, controller:, action:, params: {})
    describe "a #{verb} to #{path}" do
      before do
        @it_name = params.keys.empty? ? "routes to #{controller} #{action}" : "routes to #{controller} #{action} with parameters #{params}"
      end

      it @it_name do
        assert_routing({method: verb, path: path}, {controller: controller, action: action}.merge(params))
      end
    end
  end
```

apanzerj commented:
To Do:

- [ ] Tests
- [ ] Enforce one duplicable stage per project
- [ ] Allow duplicable setting to be set via API
- [ ] UI Pictures

sandlerr commented:
should `duplicable` have a per-project uniqueness validation?

apanzerj commented:
It will

apanzerj commented on .rubocop.yml:
```diff
@@ -1,3 +1,6 @@
+FrozenStringLiteralComment:
```
cc @grosser is this okay? Without it every file fails.

apanzerj replied:
It would also be a massive pr

apanzerj replied:
Made the massive PR here: https://github.com/zendesk/samson/pull/1200

@zendesk/samson lemme know what you want to do...

grosser replied:
push that big PR through unless there are any big blockers ...

apanzerj commented:
@zendesk/samson Can somebody give this a review?

cc @pswadi-zendesk

grosser commented on app/controllers/api/base_controller.rb:
```diff
@@ -3,6 +3,8 @@
 class Api::BaseController < ApplicationController
   skip_before_action :verify_authenticity_token
 
+  before_action :require_project
```
not really required ... `find_project` ?

apanzerj replied:
I'm just mimic-ing the concern we have. Require everywhere it's included and skip when needed.

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -18,10 +16,19 @@ def clone
     end
   end
 
+  def duplicable
```
this seems weird ...

grosser replied:
some docs on what it does / why we need that ?

apanzerj replied:
Situation:

You add a new deploy group(s) for testing. You need to know which stage is a complete stage (ie: all the commands make sense and it deploys the project successfully). Then, when you add your new stage, you simply clone the stage marked duplicable and change the deploy groups.

I wanted to call it a "Gold/Master Stage" but gold stage is meaningless and master stage was confusing.

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -18,10 +16,19 @@ def clone
+    render json: @project.duplicable_stage
+  end
+
+  def put_duplicable
```
`create_duplicate` ... how is this different from `clone` ... ? ... more docs :D

apanzerj replied:
Sorry. `#put_duplicable` is to change which stage is marked duplicable. Only one stage per project is "duplicable". The `#duplicable` method is simply the currently marked stage on the project that is duplicable. Clone is just clone.

apanzerj replied:
This changes which stage is duplicable.

grosser replied:
chose_duplicable / mark_duplicable ?

On Fri, Aug 12, 2016 at 3:52 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/api/stages_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r74671414>:
>
> > @@ -18,10 +16,19 @@ def clone
> >      end
> >    end
> >
> > +  def duplicable
> > +    render json: @project.duplicable_stage
> > +  end
> > +
> > +  def put_duplicable
>
> Sorry. #put_duplicable is to change which stage is marked duplicable.
> Only one stage per project is "duplicable". The #duplicable method is
> simply the currently marked stage on the project that is duplicable. Clone
> is just clone.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/122035a3f8635be3fd046f6620704db40db5d7c0#r74671414>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ2FRzyjFGV0ZsJblBRgHoD0LK-XAks5qfPkxgaJpZM4Jhtog>
> .
>

apanzerj replied:
mark is better.

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -18,10 +16,19 @@ def clone
 
   def stage_to_clone
-    Stage.find_by_param!(params.require(:stage_id))
+    Stage.find_by_id(params.require(:stage_id))
```
should we stay with permalinks for consistency ?

apanzerj replied:
I thought about that, but permalinks are not RESTful in my opinion. 

The permalink is a nicety, like porcelain, it is essentially an extension of the UI.

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -35,4 +42,8 @@ def deploy_group_ids
   end
+
+  def duplicable_stage
+    @project.stages.where(id: params[:id]).first
```
`first!` so we get 404s ?

apanzerj replied:
yessir.

grosser commented on app/models/stage.rb:
```diff
@@ -180,6 +180,11 @@ def deploy_group_names
   end
 
+  def duplicable!
+    project.duplicable_stage && project.duplicable_stage.update!(duplicable: false)
```
triggers 2 finds ?

apanzerj replied:
```

  def duplicable!
    old_duplicable_stage = project.duplicable_stage
    old_duplicable_stage.update!(duplicable: false) if old_duplicable_stage
    update!(duplicable: true)
  end
```

grosser replied:
Label should then also say something like "Maximum of 1 per project"

On Fri, Aug 12, 2016 at 3:57 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/models/stage.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r74671800>:
>
> > @@ -180,6 +180,11 @@ def deploy_group_names
> >      DeployGroup.enabled? ? deploy_groups.select(:name).sort_by(&:natural_order).map(&:name) : []
> >    end
> >
> > +  def duplicable!
> > +    project.duplicable_stage && project.duplicable_stage.update!(duplicable: false)
>
>
>   def duplicable!
>     old_duplicable_stage = project.duplicable_stage
>     old_duplicable_stage.update!(duplicable: false) if old_duplicable_stage
>     update!(duplicable: true)
>   end
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/122035a3f8635be3fd046f6620704db40db5d7c0#r74671800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ2oo7wQoEyn0JqLsfnjH9HLXH2Fgks5qfPpNgaJpZM4Jhtog>
> .
>

apanzerj replied:
:zap: got it

grosser commented on app/views/stages/_fields.html.erb:
```diff
@@ -61,6 +61,17 @@
+      <div class="checkbox">
+        <%= form.label :duplicable do %>
+          <%= form.check_box :duplicable %>
+          This stage can serve as a production template.
```
not sure why we need this ... if a stage goes to production it can serve as a production template ?
also I expect people will clone a stage and just leave this on, breaking it ...
... if we still want it, then pick a better column name or description ? `Can be used as template for copying / has not unique customizations`

apanzerj replied:
So here's the deal:

1. At one point in time, I cloned a stage and ran with it, not realizing it was somebody just messing around.

2. That stage had `bundle -v foo bar blah` and so it just spit out the current version of bundler. Having a "gold master" stage would fix this problem. If you can come up with a better name, I'm all ears, this was the best thing @sandlerr and I came up with.

grosser replied:
then say it "Duplication template" ... or "Clone template" ... or name the
column clone_template ... just somehow have the description and column in
sync ...

On Fri, Aug 12, 2016 at 4:01 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/views/stages/_fields.html.erb
> <https://github.com/zendesk/samson/pull/1198#discussion_r74672133>:
>
> > @@ -61,6 +61,17 @@
> >      </div>
> >    </div>
> >
> > +  <div class="form-group">
> > +    <div class="col-sm-offset-2 col-sm-10">
> > +      <div class="checkbox">
> > +        <%= form.label :duplicable do %>
> > +          <%= form.check_box :duplicable %>
> > +          This stage can serve as a production template.
>
> So here's the deal:
>
>    1.
>
>    At one point in time, I cloned a stage and ran with it, not realizing
>    it was somebody just messing around.
>    2.
>
>    That stage had bundle -v foo bar blah and so it just spit out the
>    current version of bundler. Having a "gold master" stage would fix this
>    problem. If you can come up with a better name, I'm all ears, this was the
>    best thing @sandlerr <https://github.com/sandlerr> and I came up with.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/122035a3f8635be3fd046f6620704db40db5d7c0#r74672133>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZx9HsmowCIA2_Fq2blVDvxyePgbKks5qfPtGgaJpZM4Jhtog>
> .
>

apanzerj replied:
I talked to @dragonfax and he had some good ideas, do you like any of the following:

1. Alpha stage (the progenitor stage)
2. Baseline stage
3. Prime stage

grosser commented on config/routes.rb:
```diff
@@ -3,13 +3,25 @@
 
-    resources :stages, only: [:index] do
+    resources :projects, param: :project_id, only: [] do
+      get :duplicable, on: :member, to: 'stages#duplicable'
```
use `member do` .. no need for `to` and `on`

apanzerj replied:
What's wrong with `on: :member` ?

apanzerj replied:
This needs the to: `stages#duplicable` because without it, the request routes to the projects controller.

See failure: https://travis-ci.org/zendesk/samson/jobs/151951746

grosser replied:
hmm yeah works too ... saw `member do` more often ...

On Fri, Aug 12, 2016 at 3:58 PM, Adam Panzer <notifications@github.com>
wrote:

> In config/routes.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r74671898>:
>
> >        collection do
> >          get :active_count
> >        end
> >      end
> >
> > -    resources :stages, only: [:index] do
> > +    resources :projects, param: :project_id, only: [] do
> > +      get :duplicable, on: :member, to: 'stages#duplicable'
>
> What's wrong with on: :member ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/122035a3f8635be3fd046f6620704db40db5d7c0#r74671898>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ-MUyGoEzYZUVGjUm6uo0abS6Rzpks5qfPqYgaJpZM4Jhtog>
> .
>

grosser commented on test/controllers/api/deploys_controller_test.rb:
```diff
@@ -3,7 +3,11 @@
 SingleCov.covered!
 
 describe Api::DeploysController do
+  assert_route verb: "GET", path: "/api/deploys/active_count", to: 'api/deploys#active_count'
```
not sure if these are worthwhile, the test that executes them makes sure they exist ...

apanzerj replied:
@sandlerr thought it was useful. I agree too, it helps make sure you end up with the routes you intend.

apanzerj commented:
Talked to @grosser and he thinks:

#usable_as_template?
#usable_as_template!

The route name might be:

put api/projects/1/stages/2/template
get api/projects/1/stages/2/template 

Thoughts? cc @grosser

grosser commented:
+1

On Mon, Aug 15, 2016 at 1:01 PM, Adam Panzer <notifications@github.com>
wrote:

> Talked to @grosser <https://github.com/grosser> and he thinks:
>
> #usable_as_template?
> #usable_as_template!
>
> The route name might be:
>
> put api/projects/1/stages/2/template
> get api/projects/1/stages/2/template
>
> Thoughts? cc @grosser <https://github.com/grosser>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198#issuecomment-239911787>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ_4zTog3wsPeFkjxwLOxy6Zv_l5rks5qgMWhgaJpZM4Jhtog>
> .
>

apanzerj commented:
@grosser can you review?

Also: 

So the segfault happened in the build for #0e134c5. I cleared the cache and updated webmock (where the segfault traced to) and it still happened, this time in actioncontroller journey foo.

Then I updated rails and poof, no more segfault. Could be corrupted gem cache or could be unrelated.

apanzerj commented:
cc @zendesk/samson @dragonfax and @irwaters

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,44 @@
+    render json: paginate(deploy_groups)
+  end
+
+  def project_deploy_groups
```
that should be index just with a different url ?

apanzerj replied:
So now they both go to index and look like this: 

```ruby
  def index
    scope = current_project.present? ? stage.deploy_groups : deploy_groups
    render json: paginate(scope)
  end
```

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,44 @@
+      stage.save!
+      head :no_content
+    else
+      render json: {message: "Do not modify production via the API"}, status: 403
```
wut ?

apanzerj replied:
hmm. I guess we could get rid of it. It was an attempt to protect stages from modification via the API if they are production.

grosser replied:
seems kind of random / makes the api pretty useless ...

On Tue, Aug 16, 2016 at 12:44 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/api/deploy_groups_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75005943>:
>
> > +
> > +  def index
> > +    render json: paginate(deploy_groups)
> > +  end
> > +
> > +  def project_deploy_groups
> > +    render json: paginate(stage.deploy_groups)
> > +  end
> > +
> > +  def update_deploy_groups
> > +    if !production_change?
> > +      stage.deploy_groups = deploy_groups_to_add
> > +      stage.save!
> > +      head :no_content
> > +    else
> > +      render json: {message: "Do not modify production via the API"}, status: 403
>
> hmm. I guess we could get rid of it. It was an attempt to protect stages
> from modification via the API if they are production.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/d43bdda94cc05d08fa093f7c694b09953335fe89#r75005943>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ3DV28_JWK2lyFknvwh600_UdEnQks5qghMQgaJpZM4Jhtog>
> .
>

apanzerj replied:
This was because we don't want non-humans making production changes.

grosser replied:
but it's an authorized request ... not sure why it matters how it came to
the app ... also some very arbitrary rule that makes the api useless ...

On Tue, Aug 16, 2016 at 1:10 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/api/deploy_groups_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75010128>:
>
> > +
> > +  def index
> > +    render json: paginate(deploy_groups)
> > +  end
> > +
> > +  def project_deploy_groups
> > +    render json: paginate(stage.deploy_groups)
> > +  end
> > +
> > +  def update_deploy_groups
> > +    if !production_change?
> > +      stage.deploy_groups = deploy_groups_to_add
> > +      stage.save!
> > +      head :no_content
> > +    else
> > +      render json: {message: "Do not modify production via the API"}, status: 403
>
> This was because we don't want non-humans making production changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/d43bdda94cc05d08fa093f7c694b09953335fe89#r75010128>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ3AvMCswYx1rNRuA2lzV4eQUV9dCks5qghkrgaJpZM4Jhtog>
> .
>

apanzerj replied:
I got the 'just do it' so I'll back this out.

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,44 @@
+
+  def update_deploy_groups
+    if !production_change?
+      stage.deploy_groups = deploy_groups_to_add
```
how about stage_controller#update ?

apanzerj replied:
done

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,43 @@
+  end
+
+  def search_params
+    if project_id = params.fetch(:project_id, false)
```
no need to fetch ... `if project_id = params[:project_id]`

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,43 @@
+    if project_id = params.fetch(:project_id, false)
+      stage_ids = Project.find(project_id).stages.pluck(:id)
+      search_params = {jobs: {}, deploys: {stage_id: stage_ids}}
+    elsif stage_id = params.fetch(:stage_id, false)
```
check stage_id first, since if I give project and stage then stage should win ?

apanzerj replied:
sure. In the end it's moot, there are three routes that hit this method:

api/stages/:stage_id/deploys
api/projects/:project_id/deploys
api/deploys

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,43 @@
+  protected
+
+  def filter
+    if filter_param = params.fetch(:filter, false)
```
no fetch needed

apanzerj replied:
done.

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,43 @@
+
+  def filter
+    if filter_param = params.fetch(:filter, false)
+      filter_param if ["pending", "running", "canceled", "failed", "errored", "succeeded"].include?(filter_param)
```
reuse some constant that has the list of all value states ?

grosser replied:
also should be bad_request if it is not included instead of returning all silently

apanzerj replied:
yup

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,43 @@
+    if filter
+      search_params[:jobs] = {status: filter}
+    else
+      search_params.delete(:jobs)
```
don't set jobs to empty above instead of deleting here ?

apanzerj replied:
done

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,43 @@
+    search_params
+  end
+
+  def deploy_scope
```
does very little ... maybe inline

apanzerj replied:
roger

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -19,10 +17,19 @@ def clone
   end
 
+  def template
+    render json: @project.template
```
maybe call it template_stage so it's obvious what it is

apanzerj replied:
done

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -36,4 +43,8 @@ def deploy_group_ids
   end
+
+  def new_template_stage
+    @project.stages.where(id: params[:id]).first!
```
how about `@project.stages.find(params[:id])`

apanzerj replied:
done

grosser commented on app/controllers/concerns/deploy_groups_common.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+module DeployGroupsCommon
+  extend ActiveSupport::Concern
```
don't need concern ?

apanzerj replied:
in this case, no, but I might need it later?

Take it out?

grosser replied:
take out anything we don't need :)

On Tue, Aug 16, 2016 at 2:00 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/concerns/deploy_groups_common.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75019220>:
>
> > @@ -0,0 +1,8 @@
> > +# frozen_string_literal: true
> > +module DeployGroupsCommon
> > +  extend ActiveSupport::Concern
>
> in this case, no, but I might need it later?
>
> Take it out?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/d43bdda94cc05d08fa093f7c694b09953335fe89#r75019220>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZyX6PtX8MiIzN6Du_8L8KIb0Mxd-ks5qgiT9gaJpZM4Jhtog>
> .
>

grosser replied:
take it out :D

apanzerj replied:
out.

grosser commented on app/controllers/stages_controller.rb:
```diff
@@ -102,4 +99,12 @@ def stage_params
+  end
+
+  def update_template
+    template_param ? @stage.project.template!(@stage) : @stage.project.template!
```
make this a after_save hook ?

not sure what template! does ... maybe pick a more telling name ?

grosser replied:
... remove default and call with `template!(nil)` ... much more obvious

grosser replied:
this means not selection template will unset the projects template ... seems broken

apanzerj replied:
eh?

template!(nil) => Just unset the template stage for the project.
template!(<some stage>) => Unset the existing template for the project (if there is one) and set the one given.

apanzerj replied:
Also: this can't be an after_save hook because the default behavior will still set the passed in stage as template w/o unsetting the previous stage.

grosser replied:
can you make that more intuitive ? maybe keyword args ?

On Tue, Aug 16, 2016 at 1:59 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/stages_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75018949>:
>
> > @@ -102,4 +99,12 @@ def stage_params
> >    def find_stage
> >      @stage = current_project.stages.find_by_param!(params[:id])
> >    end
> > +
> > +  def template_param
> > +    @template_param ||= ActiveRecord::Type::Boolean.new.type_cast_from_database(stage_params.delete("template"))
> > +  end
> > +
> > +  def update_template
> > +    template_param ? @stage.project.template!(@stage) : @stage.project.template!
>
> eh?
>
> template!(nil) => Just unset the template stage for the project.
> template!() => Unset the existing template for the project (if there is
> one) and set the one given.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/d43bdda94cc05d08fa093f7c694b09953335fe89#r75018949>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ6zi1cgXp8DRM4SvuCDFOyMErZL8ks5qgiSmgaJpZM4Jhtog>
> .
>

grosser replied:
why can't it unset the previous stage when `template` is enabled ?

On Tue, Aug 16, 2016 at 2:00 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/stages_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75019104>:
>
> > @@ -102,4 +99,12 @@ def stage_params
> >    def find_stage
> >      @stage = current_project.stages.find_by_param!(params[:id])
> >    end
> > +
> > +  def template_param
> > +    @template_param ||= ActiveRecord::Type::Boolean.new.type_cast_from_database(stage_params.delete("template"))
> > +  end
> > +
> > +  def update_template
> > +    template_param ? @stage.project.template!(@stage) : @stage.project.template!
>
> Also: this can't be an after_save hook because the default behavior will
> still set the passed in stage as template w/o unsetting the previous stage.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/d43bdda94cc05d08fa093f7c694b09953335fe89#r75019104>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ9Yk8WIjCIV9Yz09hIjOjewZz8lQks5qgiTdgaJpZM4Jhtog>
> .
>

grosser replied:
any answer for this ?

apanzerj replied:
sorry. mega pr. 

I'll add kwargs. 

There's two modes because, if you simply want to unset the current template, you need a way to do that without specifying a new stage as template.

grosser commented on app/models/project.rb:
```diff
@@ -43,6 +43,18 @@ class Project < ActiveRecord::Base
   scope :search, ->(name) { where("name like ?", "%#{name}%") }
 
+  def template
+    stages.where(template: true)&.first
```
don't need the `&`

apanzerj replied:
done.

grosser commented on app/views/stages/_fields.html.erb:
```diff
@@ -61,6 +61,17 @@
+      <div class="checkbox">
+        <%= form.label :template do %>
+          <%= form.check_box :template %>
+          This stage can serve as a production template (maximum of 1 per project).
```
how about: `Can be cloned and modified to create a new stage (1 per project)`

apanzerj replied:
ehhh. All the stages can be cloned. The point about this is that this particular stage *should* be cloned.

grosser commented on config/routes.rb:
```diff
@@ -4,13 +4,34 @@
+
+    resources :projects, param: :project_id, only: [] do
+      member do
+        get :template, to: 'stages#template'
```
make it a collection on stages and do `project/foo/stages/template` ?

grosser replied:
hmmm I guess that could break permalinks ...

apanzerj replied:
eh? there are no permalinks in the api...

apanzerj replied:
but I like the idea of the route being ```project/:id/stages/template```

grosser replied:
should be fine too ... chances are nobody named their stage template :D

On Tue, Aug 16, 2016 at 1:38 PM, Adam Panzer <notifications@github.com>
wrote:

> In config/routes.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75015059>:
>
> >        collection do
> >          get :active_count
> >        end
> >      end
> >
> > -    resources :stages, only: [:index] do
> > +    resources :deploy_groups, only: [:index]
> > +
> > +    resources :projects, param: :project_id, only: [] do
> > +      member do
> > +        get :template, to: 'stages#template'
>
> but I like the idea of the route being project/:id/stages/template
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/d43bdda94cc05d08fa093f7c694b09953335fe89#r75015059>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ0TeJzaVABU6DV1TAVwW3B9AkfnVks5qgh_HgaJpZM4Jhtog>
> .
>

apanzerj replied:
even if they did, it wouldn't be project/id/stages/template, it would be project/id/stages/stage_id

grosser commented on config/routes.rb:
```diff
@@ -4,13 +4,34 @@
+      resources :stages, only: [:index] do
+        member do
+          put :template, to: 'stages#mark_template'
+          get :deploy_groups, to: 'deploy_groups#project_deploy_groups'
```
include deploy groups in normal stage#get ?

apanzerj replied:
This is now moved per previous comment. Still relevant?

grosser commented on config/routes.rb:
```diff
@@ -4,13 +4,34 @@
+    end
+
+    resources :stages, only: [] do
+      get :deploys, to: 'deploys#index'
```
what url is this ? stages/deploys ?

grosser replied:
or same as `resources :deploys, only :index` ?

apanzerj replied:
```api_stage_deploys GET    /api/stages/:stage_id/deploys(.:format)                          api/deploys#index```

apanzerj replied:
In order to get the last successful deploy for a given stage.

grosser commented on test/controllers/api/deploy_groups_controller_test.rb:
```diff
@@ -0,0 +1,155 @@
+
+  describe 'when deploy_groups is disabled' do
+    before do
+      DeployGroup.stubs(:enabled?).returns(false)
```
one of these stubs is the default and is not needed

apanzerj replied:
⚡

grosser commented on test/controllers/api/deploy_groups_controller_test.rb:
```diff
@@ -0,0 +1,155 @@
+      assert_response :precondition_failed
+    end
+  end
+
```
I'd prefer having 1 `it 'returns precondition failed when deploy groups are disabled'` instead of wrapping everything in another layer of describes

apanzerj replied:
got it.

grosser commented on test/controllers/api/deploys_controller_test.rb:
```diff
@@ -4,7 +4,13 @@
 SingleCov.covered!
 
 describe Api::DeploysController do
+  Maxitest.timeout = false
```
this is a global change ... revert

apanzerj replied:
yup, doing so now.

grosser commented on test/controllers/api/deploys_controller_test.rb:
```diff
@@ -25,52 +31,117 @@
+      end
+
+      describe 'if the route is a project' do
+        subject do
```
modify the params of the controller instead of overloading subject ?

apanzerj replied:
done

apanzerj commented:
@grosser ready when you are

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+class Api::DeployGroupsController < Api::BaseController
+  include DeployGroupsCommon
+  before_action :enabled?
```
not a boolean method -> `ensure_enabled` or something like that

apanzerj replied:
right

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  before_action :enabled?
+
+  def index
+    scope = current_project.present? ? stage.deploy_groups : deploy_groups
```
no need for `.present?`

grosser replied:
missing `.sort_by(&:natural_order)` for the stage side ... maybe best to copy that bit and not have a concern ?

apanzerj replied:
so just nix the concern?

grosser replied:
yeah

On Tue, Aug 16, 2016 at 5:15 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/api/deploy_groups_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75044310>:
>
> > @@ -0,0 +1,22 @@
> > +# frozen_string_literal: true
> > +class Api::DeployGroupsController < Api::BaseController
> > +  include DeployGroupsCommon
> > +  before_action :enabled?
> > +
> > +  def index
> > +    scope = current_project.present? ? stage.deploy_groups : deploy_groups
>
> so just nix the concern?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/66abc156d3f3ad8478ca07ee5c0e76ed494978bf#r75044310>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ7mfyXv4UAHURB7-DypoHmNfjbMZks5qglKwgaJpZM4Jhtog>
> .
>

apanzerj replied:
taken out.

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  protected
+
+  def stage
+    current_project.stages.where(id: params[:id]).first!
```
`current_project.stages.find(params[:id])`

apanzerj replied:
changed to: `current_project.stages.find(params[:id])`

grosser commented on app/controllers/api/deploy_groups_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  def enabled?
+    return if DeployGroup.enabled?
+    render json: {message: "DeployGroups are not enabled."}, status: :precondition_failed
+    false
```
no need for `false`

apanzerj replied:
roger

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,40 @@
+
+  protected
+
+  def filter
```
rename to job_filter ?

apanzerj replied:
sure

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,40 @@
+    search_params
+  end
+
+  def ensure_filter
```
no need for true or false

apanzerj replied:
zapped

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,40 @@
+  end
+
+  def ensure_filter
+    return true if params[:filter].nil?
```
`unless params[:filter]`

apanzerj replied:
yup

grosser commented on app/controllers/api/deploys_controller.rb:
```diff
@@ -1,16 +1,40 @@
+  def ensure_filter
+    return true if params[:filter].nil?
+    unless Job.valid_status?(filter)
+      render json: { error: "Filter is not valid. Please use " + Job::VALID_STATUSES.join(", ") }, status: :bad_request
```
nice!

grosser commented on app/controllers/api/projects_controller.rb:
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+class Api::ProjectsController < Api::BaseController
+  def index
+    render json: Project.ordered_for_user(current_user).all
```
does this use the project_serializer or does that render out all attributes ?

apanzerj replied:
I'll write a test and let you know

apanzerj replied:
According to test: yes

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -19,10 +18,25 @@ def clone
+  end
+
+  def template_stage
+    render json: current_project.template
```
404 when there is no template ?

apanzerj replied:
404 will raise in an api client, so we don't necessarily want to do that... right?

grosser replied:
I think we want that / it's the normal response for 'not found' :D

On Tue, Aug 16, 2016 at 5:28 PM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/api/stages_controller.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75045355>:
>
> > @@ -19,10 +18,25 @@ def clone
> >      end
> >    end
> >
> > +  def update
> > +    stage.deploy_groups = deploy_groups
> > +    stage.save!
> > +    head :no_content
> > +  end
> > +
> > +  def template_stage
> > +    render json: current_project.template
>
> 404 will raise in an api client, so we don't necessarily want to do
> that... right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/66abc156d3f3ad8478ca07ee5c0e76ed494978bf#r75045355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ2Kz21UUA9sCrI9gKXW_8h7-cSBlks5qglWRgaJpZM4Jhtog>
> .
>

apanzerj replied:
done.

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -1,11 +1,10 @@
-
-  before_action :require_project
+  skip_before_action :require_project, only: :clone
+  before_action :production_change!, if: :production_change?, only: :update
```
-> `ensure_non_production` and then inline the checker logic ? ... easier to read and less clutter here ...

apanzerj replied:
zapping the logic. was told it was overkill

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -19,10 +18,25 @@ def clone
   end
 
+  def update
+    stage.deploy_groups = deploy_groups
```
want to accept all params just like stages_controller ?

grosser replied:
... good enough for now, but I'd expect someone wants to do more then deploy_groups :D

grosser commented on app/controllers/api/stages_controller.rb:
```diff
@@ -36,4 +50,17 @@ def deploy_group_ids
   def deploy_groups
```
params should not be deploy_group_ids, but params[:stage][:deploy_group_ids] -> open for future additions ?

apanzerj replied:
Fixed. New params structure is: 

{ stage: { source_stage_id: , name: , deploy_group_ids: } }

grosser commented on app/models/project.rb:
```diff
@@ -43,6 +43,18 @@ class Project < ActiveRecord::Base
 
   scope :search, ->(name) { where("name like ?", "%#{name}%") }
 
+  def template
```
can make this a has_one -> get caching from rails ... also name it `template_stage` ?

apanzerj replied:
Do you mean:

```
class Project < ActiveRecord::Base
  has_one: :template_stage
end

class TemplateStage < Stage
end
```

grosser replied:
something like:

has_one :template_stage, class_name: 'Stage', -> { where(template: true) }

On Wed, Aug 17, 2016 at 9:59 AM, Adam Panzer <notifications@github.com>
wrote:

> In app/models/project.rb
> <https://github.com/zendesk/samson/pull/1198#discussion_r75162775>:
>
> > @@ -43,6 +43,18 @@ class Project < ActiveRecord::Base
> >
> >    scope :search, ->(name) { where("name like ?", "%#{name}%") }
> >
> > +  def template
>
> Do you mean:
>
> class Project < ActiveRecord::Base
>   has_one: :template_stage
> end
>
> class TemplateStage < Stage
> end
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198/files/66abc156d3f3ad8478ca07ee5c0e76ed494978bf#r75162775>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ3G6NOvls5jcWEiYbQQnWJ8CeY__ks5qgz3ygaJpZM4Jhtog>
> .
>

jonmoter commented:
@apanzerj could you try to break this work into smaller PRs? This PR now has 48 files changes and 146 comments.

If you made each of those 5 bullet points in the description a separate PR, it would be way easier to manage this.

jonmoter commented:
Or at least have future PRs be more targeted?

apanzerj commented:
Yeah. This one got away from me. Sorry!

On Wed, Aug 17, 2016 at 1:04 PM, Jon Moter <notifications@github.com> wrote:

> Or at least have future PRs be more targeted?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1198#issuecomment-240530442>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABoVy9YDU7Udzf_ZFlA2f92qGAzj5jWIks5qg2lRgaJpZM4Jhtog>
> .
>

apanzerj commented:
This last change should get me a green build. I can try to break this up
once that goes through.

On Wed, Aug 17, 2016 at 1:05 PM, Adam Panzer <apanzer@zendesk.com> wrote:

> Yeah. This one got away from me. Sorry!
>
> On Wed, Aug 17, 2016 at 1:04 PM, Jon Moter <notifications@github.com>
> wrote:
>
>> Or at least have future PRs be more targeted?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/zendesk/samson/pull/1198#issuecomment-240530442>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ABoVy9YDU7Udzf_ZFlA2f92qGAzj5jWIks5qg2lRgaJpZM4Jhtog>
>> .
>>
>
>

apanzerj commented:
@jonmoter @grosser last one!

grosser commented on config/initializers/statsd.rb:
```diff
@@ -31,7 +31,7 @@ class << Samson
 Samson.statsd.namespace = "samson.app"
 
-Samson.statsd.event "Startup", "Samson startup" if ENV['SERVER_MODE']
+Samson.statsd.event "Startup".dup, "Samson startup" if ENV['SERVER_MODE']
```
no longer need this after statsd update

grosser commented on test/controllers/api/stages_controller_test.rb:
```diff
@@ -26,28 +38,8 @@
-      end
-    end
-
     it 'renders a cloned stage' do
```
`it "clones a stage"` ?

grosser commented on app/controllers/stages_controller.rb:
```diff
@@ -102,4 +99,12 @@ def stage_params
+
+  def update_template
+    template_param ? @stage.project.template_stage!(@stage) : @stage.project.reset_template_stage!
+  end
```
this still looks wrong to me:
 - removing the template param -> remove the attribute -> simple assignment
 - not changing then template param -> no change ... bug: atm it unsets the template
 - adding template param -> set the attribute -> can be a assignment + after-save remove it from all previous stage

apanzerj replied:
Yeah. I see what you're saying.

grosser commented on test/controllers/stages_controller_test.rb:
```diff
@@ -221,13 +234,29 @@
         end
 
         describe 'valid attributes' do
+          [[true, "0"], [false, "1"]].each do |bool, val|
```
if you submit as true/false this tests gets a bit simpler and the code makes more sense :D
